api/crypto/aes.go

```diff
@@ -1,26 +1,60 @@
 package crypto
 
 import (
+	"bytes"
 	"crypto/aes"
 	"crypto/cipher"
+	"crypto/rand"
 	"io"
 
 	"golang.org/x/crypto/scrypt"
 )
 
-// NOTE: has to go with what is considered to be a simplistic in that it omits any
-// authentication of the encrypted data.
-// Person with better knowledge is welcomed to improve it.
-// sourced from https://golang.org/src/crypto/cipher/example_test.go
+// AesEncrypt reads from input, encrypts with AES-256 and writes to output. passphrase is used to generate an encryption key.
+// The encryption uses AES-256 in CTR mode.
+// The decryption function supports the previous portainer backups which use OFB mode.  New archives
+// will use CTR mode and have a header to indicate the encryption mode.
 
-var emptySalt []byte = make([]byte, 0)
+// The encrypted file header
+const ctrHeader = "AES256-CTR"
 
-// AesEncrypt reads from input, encrypts with AES-256 and writes to the output.
-// passphrase is used to generate an encryption key.
 func AesEncrypt(input io.Reader, output io.Writer, passphrase []byte) error {
-	// making a 32 bytes key that would correspond to AES-256
-	// don't necessarily need a salt, so just kept in empty
-	key, err := scrypt.Key(passphrase, emptySalt, 32768, 8, 1, 32)
+	return AesEncryptCTR(input, output, passphrase)
+}
+
+func AesDecrypt(input io.Reader, passphrase []byte) (io.Reader, error) {
+	// Read file metadata to determine how it was encrypted
+	var buf bytes.Buffer
+	tee := io.TeeReader(input, &buf)
```

Comment:
This could be a problem if the backup is too big, as it will waste a lot more memory, which may not be available.

Maybe consider reading the header from `input` and passing it directly to `AesDecryptCTR()`, while passing an `io.MultiReader()` to `AesDecryptOFB()` composed of the first chunk you read for the header and then the remaning `input`.

A cleaner alternative could be to use a `bufio.Reader` to wrap `input` so you can `Peek()` for the header without having to advance the reader.